### PR TITLE
Unit tests

### PR DESCRIPTION
**Pull request by marcin-lawrowski**
WIP

### CODE REVIEW

**Comment by marcin-lawrowski** [comment]
It's a WIP

File: includes/render.php [hunk 1]
```diff
@@ -61,7 +61,7 @@ class GEM_Form_Fields {
 
 	private static $cycle = 0;
 
-	public function dispatch_field( $field, $cycle = 1 ) {
+	public static function dispatch_field( $field, $cycle = 1 ) {
```
Comment (lukecarbis) [inline includes/render.php]: Do these need to be static? Would it be better to change the way they're called, instead of making them static?

Comment (lukecarbis) [inline includes/render.php]: (I have an aversion to unnecessary static functions)

File: tests/mock-transport.php [hunk 2]
```diff
@@ -0,0 +1,20 @@
+<?php
+
+class WP_Http_Mock_Transport {
+
+	public static $response = array();
+	public static $test_class;
+	public static $expected_url;
+
+	public static function test() {
+		return true;
```
Comment (lukecarbis) [inline tests/mock-transport.php]: What's this for?

File: tests/test-class-gem-dispatcher.php [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+	/**
+	 * PHP unit setup function
+	 *
+	 * @return void
+	 */
+	function setUp() {
+		parent::setUp();
+
+		\WP_Http_Mock_Transport::$test_class = $this;
+		add_action( 'http_api_transports', array( $this, 'get_transports' ) );
```
Comment (lukecarbis) [inline tests/test-class-gem-dispatcher.php]: Nice. Very nice. 🙌

File: includes/render.php (hunk 1 above)
Comment (marcin-lawrowski) [inline includes/render.php]: They are called statically, for example:
GEM_Form_Fields::dispatch_field( $field, self::$loops )

File: tests/test-class-gem-dispatcher.php [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+		$sample_data = 'the_sample';
+		$user_name = 'the_user';
+		$sample_response = '{ "result": "OK" }';
+
+		update_option( 'gem-settings', false );
+		$this->assertFalse( \GEM_Dispatcher::get_user_level() );
+
+		update_option( 'gem-settings', array( 'username' => $user_name ) );
+		set_transient( 'gem-' . $user_name . '-account', $sample_data );
+		$this->assertEquals( $sample_data, \GEM_Dispatcher::get_user_level( ) );
```
Comment (lukecarbis) [inline tests/test-class-gem-dispatcher.php]: No space between brackets.

File: tests/test-class-gem-dispatcher.php [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+		delete_transient( 'gem-' . $user_name . '-account' );
+
+		\WP_Http_Mock_Transport::$expected_url = "http://api.madmimi.com/user/account_status?username=$user_name";
+		\WP_Http_Mock_Transport::$response = array(
+			'response' => array(
+				'code' => 401,
+			),
+		);
+		$this->assertFalse( \GEM_Dispatcher::get_user_level( ) );
```
Comment (lukecarbis) [inline tests/test-class-gem-dispatcher.php]: No space between brackets.

File: tests/test-class-gem-dispatcher.php [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+
+		\WP_Http_Mock_Transport::$expected_url = "http://api.madmimi.com/user/account_status?username=$user_name";
+		\WP_Http_Mock_Transport::$response = array(
+			'response' => array(
+				'code' => 200,
+			),
+			'body' => $sample_response,
+		);
+		$this->assertEquals( 'OK', \GEM_Dispatcher::get_user_level( ) );
```
Comment (lukecarbis) [inline tests/test-class-gem-dispatcher.php]: No space between brackets.

File: tests/test-class-gem-dispatcher.php [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+	public function test_user_sign_in() {
+		$sample_response = 'the_response';
+
+		\WP_Http_Mock_Transport::$expected_url = 'http://api.madmimi.com/sessions/single_signon_token';
+		\WP_Http_Mock_Transport::$response = array(
+			'response' => array(
+				'code' => 401,
+			),
+		);
+		$this->assertFalse( \GEM_Dispatcher::user_sign_in( ) );
```
Comment (lukecarbis) [inline tests/test-class-gem-dispatcher.php]: No space between brackets.

File: tests/test-class-gem-dispatcher.php [hunk 8]
```diff
@@ -0,0 +1,224 @@
...
+
+		\WP_Http_Mock_Transport::$expected_url = 'http://api.madmimi.com/sessions/single_signon_token';
+		\WP_Http_Mock_Transport::$response = array(
+			'response' => array(
+				'code' => 200,
+			),
+			'body' => $sample_response,
+		);
+		$this->assertEquals( 'http://api.madmimi.com/sessions/single_signon?token=' . $sample_response, \GEM_Dispatcher::user_sign_in( ) );
```
Comment (lukecarbis) [inline tests/test-class-gem-dispatcher.php]: No space between brackets.

File: tests/mock-transport.php (hunk 2 above)
Comment (marcin-lawrowski) [inline tests/mock-transport.php]: WordPress calls this method in order to determine if the transport is ready. Simply - it does not work without the method.

**Comment by lukecarbis** [comment]
This all looks great so far @marcin-lawrowski!

File: tests/mock-transport.php (hunk 2 above)
Comment (lukecarbis) [inline tests/mock-transport.php]: :+1:

Comment (lukecarbis) [inline tests/mock-transport.php]: Maybe add a comment to that effect?

File: includes/render.php (hunk 1 above)
Comment (lukecarbis) [inline includes/render.php]: :+1: Agreed. This PR should focus on unit tests, not refactoring.

File: tests/mock-transport.php (hunk 2 above)
Comment (marcin-lawrowski) [inline tests/mock-transport.php]: Done

**Comment by marcin-lawrowski** [comment]
@lukecarbis I completed all unit tests.

**Comment by valendesigns** [comment]
Have you created a coverage report?

**Comment by marcin-lawrowski** [comment]
@valendesigns Yes, but where to publish the report? Currently the total coverage is 86.68%, "includes" directory has 95.28%

**Comment by fjarrett** [comment]
@valendesigns @marcin-lawrowski Tried adding [Codecov](https://codecov.io/github/godaddy/wp-godaddy-email-marketing) but it seems to have broken the builds. Please have a look and see what might be going wrong?

**Comment by valendesigns** [comment]
I prefer Coveralls.

**Comment by fjarrett** [comment]
@valendesigns I integrated Coveralls too. I'm not sure if any configuration for it is required since we are already using Travis and it's a public repo.

https://codecov.io/github/godaddy/wp-godaddy-email-marketing
https://coveralls.io/github/godaddy/wp-godaddy-email-marketing
https://codeclimate.com/github/godaddy/wp-godaddy-email-marketing

**Comment by valendesigns** [comment]
@fjarrett I'll take a look and see if everything is setup correctly.

**Comment by valendesigns** [comment]
@fjarrett Do you mind if I remove Codecov and integrate Coveralls?

**Comment by fjarrett** [comment]
@valendesigns Go for it.

**Comment by valendesigns** [comment]
I don't believe it will generate a coverage report right now. The commit doesn't touch any PHP files, so we need to make a small change to any PHP file and commit to generate the report.

**Comment by valendesigns** [comment]
As well, the pre-commit hook is not generating the `readme.md` from the `readme.txt`. I believe this will be solved by renaming the file to lowercase. I'll fix that and the coverage report now.

**Comment by valendesigns** [comment]
Hey Guys, coverage is now being generated and the build is passing.

**Comment by marcin-lawrowski** [comment]
Thanks @valendesigns

**Pull request merged**